### PR TITLE
feat(lsp): allow passing position to lsp.buf functions

### LINKED ISSUE #27400: Ability to pass position parameters to lsp.buf functions
### Problem

At the moment, `lsp.buf.references` [requires my cursor to be on the word for which I want to get references (calls the make_position_params function)](https://github.com/rudiejd/neovim/blob/bcf96538143564bc8de26c5db28df7cfbaf8025a/runtime/lua/vim/lsp/util.lua#L1961). If I want to request references for a symbol that isn't under my cursor, I first have to navigate there then navigate back. This also applies to `lsp.buf.implementation`, `lsp.buf.definition`, etc. 

### Expected behavior

In Lua, I should be able to pass my own [TextDocumentPositionParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentPositionParams) to my LSP server so that I can utilize all of these functions for any symbol in the buffer, not just the one under my cursor. 

Right now, my alternative options for this seem to be:
1. Do `lsp.buf_request` in lua and pass whatever parameters I want
2. Change the cursor to the position of my request, make the request, and change it back

If adding the ability to pass position isn't desired for any of the `lsp.buf` module functions, I can go with (1), but to me this seems like a feature that should be standard.

### PR DESCRIPTION
**Pull request by rudiejd**
Closes https://github.com/neovim/neovim/issues/27400 

I considered also making `window` an option that could be passed here, but that seems like it could run against the grain of this module being requests for the "current buffer"

### CODE REVIEW

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -1956,18 +1961,19 @@ local function make_position_param(window, offset_encoding)
...
 --- Creates a `TextDocumentPositionParams` object for the current buffer and cursor position.
----
+---@param row integer|nil 1-indexed row number of position, defaults to current line
+---@param col integer|nil 0-indexed column number of position, defaults to current character
 ---@param window integer|nil: window handle or 0 for current, defaults to current
 ---@param offset_encoding string|nil utf-8|utf-16|utf-32|nil defaults to `offset_encoding` of first client of buffer of `window`
 ---@return table `TextDocumentPositionParams` object
 ---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocumentPositionParams
-function M.make_position_params(window, offset_encoding)
+function M.make_position_params(row, col, window, offset_encoding)
```
Comment (wookayin) [inline runtime/lua/vim/lsp/util.lua:1970]: We should not break the function by changing the signature.

File: runtime/lua/vim/lsp/util.lua [hunk 2]
```diff
@@ -1939,10 +1939,15 @@ end
 
 ---@param window integer|nil: window handle or 0 for current, defaults to current
 ---@param offset_encoding string utf-8|utf-16|utf-32|nil defaults to `offset_encoding` of first client of buffer of `window`
-local function make_position_param(window, offset_encoding)
+---@param row integer|nil 1-indexed row number of position, defaults to current line
+---@param col integer|nil 0-indexed column number of position, defaults to current character
+---@return table position line and character number within buffer
+local function make_position_param(window, offset_encoding, row, col)
```
Comment (wookayin) [inline runtime/lua/vim/lsp/util.lua:1945]: I don't think using 1,0-indexing is a good idea for LSP APIs. see `|api-indexing|`. I think using 0-based indexing for all row and col would be better. The LSP protocol uses 0-based indexing for the Range representations.

The row, col indexing used for this function is only due to `nvim_win_get_cursor` and other mark-related APIs specifically.

File: runtime/lua/vim/lsp/buf.lua [hunk 3]
```diff
@@ -40,8 +40,12 @@ end
 
 --- Displays hover information about the symbol under the cursor in a floating
 --- window. Calling the function twice will jump into the floating window.
-function M.hover()
-  local params = util.make_position_params()
+---@param options table|nil additional options
```
Comment (wookayin) [inline runtime/lua/vim/lsp/buf.lua:43]:
```suggestion
---@param opts? table additional options
```
Not only this place.

Also whitespace is not consistent on `col`.

File: runtime/lua/vim/lsp/util.lua [hunk 4]
```diff
@@ -1939,10 +1939,15 @@ end
...
   window = window or 0
   local buf = api.nvim_win_get_buf(window)
-  local row, col = unpack(api.nvim_win_get_cursor(window))
+  if not (row or col) then
+    row, col = unpack(api.nvim_win_get_cursor(window))
+  end
```
Comment (wookayin) [inline runtime/lua/vim/lsp/util.lua:1950]: This does not align with what is documented about the default values of `row` and `col`.

**Review by wookayin — CHANGES_REQUESTED** [review]

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (rudiejd) [reply runtime/lua/vim/lsp/util.lua:1970]: d'oh, I forget the util module is public... I got annoyed by repeatedly passing `(nil, nil, row, col)`, though I suppose I could get past that by just passing a table as args

File: runtime/lua/vim/lsp/util.lua (hunk 2 above)
Comment (rudiejd) [reply runtime/lua/vim/lsp/util.lua:1945]: Gotcha, thanks for the context. I can make this 0 based for input without breaking the existing behavior

File: runtime/lua/vim/lsp/buf.lua (hunk 3 above)
Comment (rudiejd) [reply runtime/lua/vim/lsp/buf.lua:43]: I originally named this as `opts`, but wanted consistency with the arguments that other members of this module already have. Should everything be `opts?`?

File: runtime/lua/vim/lsp/util.lua (hunk 2 above)
Comment (wookayin) [reply runtime/lua/vim/lsp/util.lua:1945]: Actually.... I just realized that `buf.format()` and `make_given_range_params` is using 1,0-indexing 🤔 (which I'd want to change and break but it's difficult to change at this point). Probably wasn't carefully chosen in retrospect or there could be some reason, e.g. wanted to avoid conversion from the callsite; Some other core devs

File: runtime/lua/vim/lsp/buf.lua (hunk 3 above)
Comment (wookayin) [reply runtime/lua/vim/lsp/buf.lua:43]: I see. Options would be fine then. Either we change all or remain consistent.

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -422,6 +422,10 @@ The following changes to existing APIs or features add new behavior.
 
 • |--startuptime| reports the startup times for both processes (TUI + server) as separate sections.
 
+* |vim.lsp.start()| now maps |K| to use |vim.lsp.buf.hover()| if the server
```
Comment (MariaSolOs) [inline runtime/doc/news.txt:425]: This change looks unrelated to this PR (?).

**Review by MariaSolOs — COMMENTED** [review]
I'm not a fan of this change. I feel like we should invest in a more flexible and general solution like https://github.com/neovim/neovim/pull/22598 for LSP-handler customization.

**Comment by wookayin** [comment]
> I feel like we should invest in a more flexible and general solution like #22598 for LSP-handler customization.

I think this PR is not about LSP *handlers*, but is about adding position parameters to the "request" functions other than the current cursor location. But I can (partially) agree with you, in a sense that rather than too specific parameters for every use cases we would want to have more **general** LSP request functions that would allow injection of any additional LSP request parameters at ease. For example, we may also want to have `buf = ...`, `win = ...`, as well as cursor locations. Ideally, it would be even better if users can build such LSP parameters (e.g. [TextDocumentPositionParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentPositionParams)) *outside* using some utility APIs, and pass these tables to inject them into LSP requests being sent.

**Comment by MariaSolOs** [comment]
> I think this PR is not about LSP _handlers_, but is about adding position parameters to the "request" functions other than the current cursor location. But I can (partially) agree with you, in a sense that rather than too specific parameters for every use cases we would want to have more **general** LSP request functions that would allow injection of any additional LSP request parameters at ease. For example, we may also want to have `buf = ...`, `win = ...`, as well as cursor locations. Ideally, it would be even better if users can build such LSP parameters (e.g. [TextDocumentPositionParams](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentPositionParams)) _outside_ using some utility APIs, and pass these tables to inject them into LSP requests being sent.

Well I count this as overloading/overriding/decoration of handlers ;)

But yes, I agree with you. Perhaps expansing `vim.lsp.with` to accommodate for these scenarios would be nice, because having custom options (with several common properties) for every LSP handler doesn't feel very maintainable to me.

**Comment by rudiejd** [comment]
To me,  #22598 and `vim.lsp.with` don't seem like suitable substitutes for this. As @wookayin mentioned, this is more about changing an individual request the changing the handlers. I don't think users should have to override handlers to pass positions to them.

My use case for this would be searching for referenced of a symbol in my buffer that I'm not moused over, so the position could change every time I want to call the handler. I'd also like to be able to navigate to the definition of something that isn't under my cursor without polluting the jumplist.

I agree about the options being cumbersome here, though. Would it be helpful to have a common type for `TextDocumentPositionParams` that gets passed as an optional argument instead of an `options` argument? I originally went with the latter because some functions already had optional arguments, but the former seems more maintainable.

**Comment by mfussenegger** [comment]
I think I still don't really understand the use-case for this.

If you're programmatically making requests, I'd think you also want to deal with the result in some special way?
If that's the case we've `client.request` and `vim.lsp.buf_request_all` where you can create requests however you want. Creating a table with a couple properties is imho not that complicated that it justifies various options and various helper functions.

And if we're going to extend this in some way, we should make sure it'll compose with whatever is happening with https://github.com/neovim/neovim/issues/25509

File: runtime/lua/vim/lsp/buf.lua (hunk 3 above)
Comment (justinmk) [reply runtime/lua/vim/lsp/buf.lua:43]: > Should everything be `opts?`?

in favor of that

**Comment by justinmk** [comment]
> My use case for this would be searching for referenced of a symbol in my buffer that I'm not moused over, so the position could change every time I want to call the handler. 

Placing the cursor programmatically works there. 

> I'd also like to be able to navigate to the definition of something that isn't under my cursor without polluting the jumplist.

`:help cursor()` says "Does not change the jumplist."

In general I agree that APIs should always take a position (#25509) instead of inspecting the cursor, but these `vim.lsp.buf` APIs are mostly for "interactive" tasks (e.g. `rename()` moves the cursor).

**Comment by justinmk** [comment]
> I actually managed to accomplish what I wanted to do with `client.request`

Can you share a code sample in https://github.com/neovim/neovim/issues/27400
